### PR TITLE
Optimize info: precompute a phase-indexed table of InfoTransformer.nextFrom

### PR DESCRIPTION
**Pull request by retronym**
Fixes scala/scala-dev#543

### CODE REVIEW

File: src/compiler/scala/tools/nsc/Global.scala [hunk 1]
```diff
@@ -1187,8 +1187,11 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
       }
       // Create phases and link them together. We supply the previous, and the ctor sets prev.next.
       val last  = components.foldLeft(NoPhase: Phase)((prev, c) => c newPhase prev)
-      // rewind (Iterator.iterate(last)(_.prev) dropWhile (_.prev ne NoPhase)).next
-      val first = { var p = last ; while (p.prev ne NoPhase) p = p.prev ; p }
+      val phaseList = Iterator.iterate(last)(_.prev).takeWhile(_ != NoPhase).toList.reverse
+      val minId = phaseList.map(_.id).min
+      val maxId = phaseList.map(_.id).max
+      nextFrom = Array.tabulate(phaseList.map(_.id).max)(i => infoTransformers.nextFrom(i))
```
Comment (mkeskells) [inline src/compiler/scala/tools/nsc/Global.scala]: Array.tabulate(maxId)

File: src/compiler/scala/tools/nsc/Global.scala [hunk 2]
```diff
@@ -1187,8 +1187,11 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
       }
       // Create phases and link them together. We supply the previous, and the ctor sets prev.next.
       val last  = components.foldLeft(NoPhase: Phase)((prev, c) => c newPhase prev)
-      // rewind (Iterator.iterate(last)(_.prev) dropWhile (_.prev ne NoPhase)).next
-      val first = { var p = last ; while (p.prev ne NoPhase) p = p.prev ; p }
+      val phaseList = Iterator.iterate(last)(_.prev).takeWhile(_ != NoPhase).toList.reverse
+      val minId = phaseList.map(_.id).min
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/Global.scala]: Seems unused.

File: src/compiler/scala/tools/nsc/Global.scala [hunk 3]
```diff
@@ -1187,8 +1187,11 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
...
+      val maxId = phaseList.map(_.id).max
+      nextFrom = Array.tabulate(maxId)(i => infoTransformers.nextFrom(i))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/Global.scala:1212]: I haven't looked how ids are allocated, but maybe we should make sure they are not (in some future) assigned randomly :-)

**Review by adriaanm — CHANGES_REQUESTED** [review]
Apologies for delay -- lost track of this one. Small nit for an unused variable, and a quick check / doc comment request.

File: src/compiler/scala/tools/nsc/Global.scala (hunk 3 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/Global.scala:1212]: It's just `val id: Id = if (prev eq null) 0 else prev.id + 1`, which seems unlikely to drift.

**Review by adriaanm — APPROVED** [review]

**Pull request merged**